src/data/types.rs

```diff
@@ -281,113 +281,165 @@ pub(super) fn print_type(
-                None => lower.as_str(),
-            };
-            write!(f, "{}{}", if *signed { "" } else { "unsigned " }, substr)
+fn write_struct_type(struct_type: &StructType, f: &mut Formatter) -> fmt::Result {
+    match struct_type {
+        StructType::Named(name, _) => {
+            write!(f, "{}", name)?;
+        }
+        StructType::Anonymous(members) => {
+            write!(f, "{{ ")?;
+            for member in members.iter() {
+                write!(f, "{};", member)?;
+            }
+            write!(f, " }}")?;
         }
-        Bool => write!(f, "_Bool"),
-        Float | Double | Void => write!(f, "{}", format!("{:?}", ctype).to_lowercase()),
-        Pointer(inner, _) | Array(inner, _) => print_pre(inner, f),
-        Function(ftype) => print_type(&ftype.return_type, None, f),
-        Enum(Some(ident), _) => write!(f, "enum {}", ident),
-        Enum(None, _) => write!(f, "<anonymous enum>"),
-        Union(StructType::Named(ident, _)) => write!(f, "union {}", ident),
-        Union(_) => write!(f, "<anonymous union>"),
-        Struct(StructType::Named(ident, _)) => write!(f, "struct {}", ident),
-        Struct(_) => write!(f, "<anonymous struct>"),
-        VaList => write!(f, "va_list"),
-        Error => write!(f, "<type error>"),
     }
+    Ok(())
 }
 
-fn print_mid(ctype: &Type, name: Option<InternedStr>, f: &mut Formatter) -> fmt::Result {
-    match ctype {
-        Type::Pointer(to, qs) => {
-            let name = name.unwrap_or_default();
-            // what do we do for (**p)()?
-            // we have to look arbitrarily deep into the type,
-            // but also we have to only print the ( once,
-            // so how do we know we know if it's already been printed?
-            let depth = match &**to {
-                Type::Array(_, _) | Type::Function(_) => true,
-                _ => false,
+pub fn print_declarator(ctype: &Type, name: Option<InternedStr>, f: &mut Formatter) -> fmt::Result {
+    fn unroll_type(ctype: &Type) -> Vec<&Type> {
+        let mut types = Vec::new();
+        let mut next_type = ctype;
+        loop {
+            types.push(next_type);
+            next_type = match next_type {
+                Type::Array(of, _) => of.as_ref(),
+                Type::Pointer(to, _) => to.as_ref(),
+                Type::Function(FunctionType { return_type, .. }) => return_type.as_ref(),
+                _ => break,
             };
-            print_mid(to, None, f)?;
+        }
+        types
+    }
+
+    use std::fmt::Write;
+    use Type::*;
+
+    let unrolled_type = unroll_type(ctype);
 
-            write!(f, " ")?;
-            if depth {
-                write!(f, "(")?;
+    let mut prefixes = Vec::new();
+    let mut postfixes = Vec::new();
+
+    // Need to skip the last item because that's the final type that needs to be
+    // put in as the specifier
+    for (index, declarator_type) in unrolled_type[..unrolled_type.len() - 1].iter().enumerate() {
```

Comment:
Maybe this would be nicer using [`windows(2)`](https://doc.rust-lang.org/std/primitive.slice.html#method.windows)? That's mutually incompatible with my suggestion to use `pop()` though :laughing:

Reply:
I also considered windows, you can't use windows with a for loop, so I figure this would be nicer.

Reply:
I don't see why not? `Windows` is also an iterator: https://doc.rust-lang.org/std/slice/struct.Windows.html#impl-ExactSizeIterator